pilot/pkg/security/authn/utils/utils.go

```diff
@@ -30,15 +30,17 @@ import (
 )
 
 // BuildInboundFilterChain returns the filter chain(s) corresponding to the mTLS mode.
-func BuildInboundFilterChain(mTLSMode model.MutualTLSMode, sdsUdsPath string, node *model.Proxy) []networking.FilterChain {
+func BuildInboundFilterChain(mTLSMode model.MutualTLSMode, sdsUdsPath string, node *model.Proxy,
+	listenerProtocol networking.ListenerProtocol) []networking.FilterChain {
 	if mTLSMode == model.MTLSDisable || mTLSMode == model.MTLSUnknown {
 		return nil
 	}
 
 	meta := node.Metadata
 	var alpnIstioMatch *ldsv2.FilterChainMatch
 	var tls *auth.DownstreamTlsContext
-	if util.IsTCPMetadataExchangeEnabled(node) {
+	if util.IsTCPMetadataExchangeEnabled(node) &&
+		(listenerProtocol == networking.ListenerProtocolTCP || listenerProtocol == networking.ListenerProtocolAuto) {
```

Comment:
what's the intended behavior here? if protocol is auto, which means protocol sniffing? and we always adds MX filter, even if the sniffed results are http? If that's fine, why we adding this to Protocol HTTP would be problematic?

Reply:
We are adding for protocol sniffing because at this point we don't know what would be the result of sniffing, if it would be http or tcp filterchain and hence we add the protocol and remove it later when we are building the filterchain in listener.go when we are sure this filterchain is for http...